examples/research_projects/vqgan-clip/VQGAN_CLIP.py

```diff
@@ -0,0 +1,266 @@
+            else:
+                yield vector
+
+    def _init_logging(self, positive_prompts, negative_prompts, image_path):
+        wandb.init(reinit=True, project="face-editor")
+        wandb.config.update({"Positive Prompts": positive_prompts})
+        wandb.config.update({"Negative Prompts": negative_prompts})
+        wandb.config.update(dict(lr=self.lr, iterations=self.iterations))
+        if image_path:
+            image = Image.open(image_path)
+            image = image.resize((256, 256))
+            wandb.log("Original Image", wandb.Image(image))
+
+    def process_prompts(self, prompts):
+        if not prompts:
+            return []
+        processed_prompts = []
+        weights = []
+        if isinstance(prompts, str):
+            prompts = [prompt.strip() for prompt in prompts.split("|")]
+        for prompt in prompts:
+            if isinstance(prompt, (tuple, list)):
+                processed_prompt = prompt[0]
+                weight = float(prompt[1])
+            elif ":" in prompt:
+                processed_prompt, weight = prompt.split(":")
+                weight = float(weight)
+            else:
+                processed_prompt = prompt
+                weight = 1.0
+            processed_prompts.append(processed_prompt)
+            weights.append(weight)
+        return {
+            "prompts": processed_prompts,
+            "weights": torch.tensor(weights, device=self.device),
+        }
+
+    def generate(
+        self,
+        pos_prompts,
+        neg_prompts=None,
+        image_path=None,
+        show_intermediate=True,
+        save_intermediate=False,
+        show_final=True,
+        save_final=True,
+        save_path=None,
+    ):
+        """Generate an image from the given prompts.
+        If image_path is provided, the image is used as a starting point for the optimization.
+        If image_path is not provided, a random latent vector is used as a starting point.
+        You must provide at least one positive prompt, and optionally provide negative prompts.
+        Prompts must be formatted in one of the following ways:
+        - A single prompt as a string, e.g "A smiling woman"
+        - A set of prompts separated by pipes: "A smiling woman | a woman with brown hair"
+        - A set of prompts and their weights separated by colons: "A smiling woman:1 | a woman with brown hair: 3" (default weight is 1)
+        - A list of prompts, e.g ["A smiling woman", "a woman with brown hair"]
+        - A list of prompts and weights, e.g [("A smiling woman", 1), ("a woman with brown hair", 3)]
+        """
+        if image_path:
+            self.latent = self._get_latent(image_path)
+        else:
+            self.latent = torch.randn(self.latent_dim, device=self.device)
+        if self.log:
+            self._init_logging(pos_prompts, neg_prompts, image_path)
+
+        assert pos_prompts, "You must provide at least one positive prompt."
```

Comment:
This should be an exception as it's checking inputs - asserts are there for checking bugs in the code 🐛